src/nimblepkg/publish.nim

```diff
@@ -0,0 +1,191 @@
+# Copyright (C) Andreas Rumpf. All rights reserved.
+# BSD License. Look at license.txt for more info.
+
+## Implements 'nimble publish' to create a pull request against
+## nim-lang/packages automatically.
+
+import httpclient, base64, strutils, rdstdin, json, os
+import tools, nimbletypes
+
+type
+  Auth = object
+    user: string
+    pw: string
+    token: string  ## base64 encoding of user:pw
+
+proc userAborted() =
+  raise newException(NimbleError, "User aborted the process.")
+
+proc getGithubAuth(): Auth =
+  var user = ""
+  let (output, exitCode) = doCmdEx("git config user.name")
+  if exitCode == 0:
+    user = output.string.strip
+  if user.len == 0:
+    user = readLineFromStdin("Github user name: ")
+    if user.len == 0: userAborted()
+  let pw = readPasswordFromStdin("Github password for " & user & ": ")
+  if pw.len == 0: userAborted()
+  result.user = user
+  result.pw = pw
+  result.token = encode(user & ':' & pw)
+
+proc searchFork(j: JsonNode): bool =
+  # Searches for: "fork":true recursively.
+  case j.kind
+  of JObject:
+    for k, v in items(j.fields):
+      if k == "fork" and v.kind == JBool: return v.bval
+    for k, v in items(j.fields):
+      if searchFork(v): return true
+  of JArray:
+    for x in j.elems:
+      if searchFork(x): return true
+  else: discard
+
+proc forkExists(a: Auth): bool =
+  try:
+    let x = getContent("https://api.github.com/repos/" & a.user & "/packages",
+        extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+          "Content-Type: application/x-www-form-urlencoded\c\L" &
+          "Accept: */*\c\L")
+    let j = parseJson(x)
+    result = searchFork(j)
+  except JsonParsingError, IOError:
+    result = false
+
+proc createFork(a: Auth) =
+  discard postContent("https://api.github.com/repos/nim-lang/packages/forks",
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+        "Content-Type: application/x-www-form-urlencoded\c\L" &
+        "Accept: */*\c\L")
+
+proc createPullRequest(a: Auth; packageName: string) =
+  echo "creating PR"
+  discard postContent("https://api.github.com/repos/nim-lang/packages/pulls",
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
```

Comment:
I doubt this method of auth will work if I have got two-factor authentication enabled. Could we not do oauth?